robot/xarm_env.py

```diff
@@ -1,8 +1,243 @@
-from robot.base_robot import TaskSpaceRobotEnv, LrfTaskSpaceRobotEnv
+from robot.base_robot import RobotEnv
+from typing import Any, Dict, List, Optional, Tuple
+from cam.realsense import RealSenseInterface
+from PIL import Image
+import gym
+import numpy as np
+import torch
+import torchvision.transforms as T
+XARM_SDK = '/home/xarm/Desktop/xArm-Python-SDK'
+import sys
+sys.path.append(XARM_SDK)
+from xarm.wrapper import XArmAPI
+from reward_extraction.reward_functions import RobotLearnedRewardFunction
+from robot.utils import Rate
+from r3m import load_r3m
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
 
-'''
-TODO: define
-'''
+class XArmEnv(RobotEnv):
+    def __init__(
+    self,
+    control_frequency_hz: int,
+    control_mode: str = 'default',
+    use_gripper: bool = False,
+    use_camera: bool = False,
+    use_r3m: bool = False,
+    r3m_net: torch.nn.Module = None,
+    xarm_ip: str = '192.168.1.220',
+    random_reset_home_pose: bool = False
+    ):
+        self.hz = control_frequency_hz
+        self.rate = Rate(control_frequency_hz)
+        self.control_mode = control_mode
+        self.use_gripper = use_gripper
+        self.random_reset_home_pos = random_reset_home_pose
+        self.xarm_ip = xarm_ip
+        self.robot = None
+        # TODO: This is bad, we should have a better home pos
+        if self.control_mode == 'default':
+            self.home_xyz = [553,29,435]
```

Comment:
you should get in the habit of adding units (radians, degrees, centimeters, meters) for robot pose related things. people will generally assume meters, but this is clearly not meters. things also start to get messy say if you were using realsense depth which gets returned in mm.

Reply:
unit added